core/solver/bicgstab.cpp

```diff
@@ -87,136 +88,148 @@ std::unique_ptr<LinOp> Bicgstab<ValueType>::conj_transpose() const
-            break;
-        }
-
-        // tmp = rho / prev_rho * alpha / omega
-        // p = r + tmp * (p - omega * v)
-        exec->run(bicgstab::make_step_1(r.get(), p.get(), v.get(), rho.get(),
-                                        prev_rho.get(), alpha.get(),
-                                        omega.get(), &stop_status));
-
-        get_preconditioner()->apply(p.get(), y.get());
-        system_matrix_->apply(y.get(), v.get());
-        rr->compute_dot(v.get(), beta.get());
-        // alpha = rho / beta
-        // s = r - alpha * v
-        exec->run(bicgstab::make_step_2(r.get(), s.get(), v.get(), rho.get(),
-                                        alpha.get(), beta.get(), &stop_status));
-
-        auto all_converged =
-            stop_criterion->update()
-                .num_iterations(iter)
-                .residual(s.get())
-                .implicit_sq_residual_norm(rho.get())
-                // .solution(dense_x) // outdated at this point
-                .check(RelativeStoppingId, false, &stop_status, &one_changed);
-        if (one_changed) {
-            exec->run(bicgstab::make_finalize(dense_x, y.get(), alpha.get(),
-                                              &stop_status));
-        }
-        if (all_converged) {
-            break;
-        }
-
-        get_preconditioner()->apply(s.get(), z.get());
-        system_matrix_->apply(z.get(), t.get());
-        s->compute_dot(t.get(), gamma.get());
-        t->compute_dot(t.get(), beta.get());
-        // omega = gamma / beta
-        // x = x + alpha * y + omega * z
-        // r = s - omega * t
-        exec->run(bicgstab::make_step_3(
-            dense_x, r.get(), s.get(), t.get(), y.get(), z.get(), alpha.get(),
-            beta.get(), gamma.get(), omega.get(), &stop_status));
-        swap(prev_rho, rho);
-    }
-}  // namespace solver
+    precision_dispatch_real_complex<ValueType>(
+        [&](auto dense_b, auto dense_x) {
+            using std::swap;
+            using Vector = matrix::Dense<ValueType>;
+            using AbsVector = matrix::Dense<remove_complex<ValueType>>;
+
+            constexpr uint8 RelativeStoppingId{1};
+
+            auto exec = this->get_executor();
+
+            auto one_op = initialize<Vector>({one<ValueType>()}, exec);
+            auto neg_one_op = initialize<Vector>({-one<ValueType>()}, exec);
+
+            auto r = Vector::create_with_config_of(dense_b);
+            auto z = Vector::create_with_config_of(dense_b);
+            auto y = Vector::create_with_config_of(dense_b);
+            auto v = Vector::create_with_config_of(dense_b);
+            auto s = Vector::create_with_config_of(dense_b);
+            auto t = Vector::create_with_config_of(dense_b);
+            auto p = Vector::create_with_config_of(dense_b);
+            auto rr = Vector::create_with_config_of(dense_b);
+
+            auto alpha =
+                Vector::create(exec, dim<2>{1, dense_b->get_size()[1]});
```

Comment:
Same issue here, but at least alpha is used.

To further state the potential issue, both beta and rho at least are used in `compute_dot` which could potentially be overriden by users as well, so should we consider these variables as "interface level" variables and thus they should have a polymorphic behavior?